quickwit/quickwit-search/src/leaf.rs

```diff
@@ -1000,134 +995,236 @@ impl CanSplitDoBetter {
-        match self {
-            CanSplitDoBetter::SplitIdHigher(_) => {
-                splits.sort_unstable_by(|a, b| b.split_id.cmp(&a.split_id))
+    /// If there's a time range and not enough splits contain at least the number of requested
+    /// documents, return None.
+    fn get_min_required_splits(
+        splits: &[SplitIdAndFooterOffsets],
+        request: &SearchRequest,
+    ) -> Option<usize> {
+        let num_requested_docs = request.start_offset + request.max_hits;
+
+        let mut min_required_splits = 0;
+        let mut partial_sum = 0;
+
+        for split in splits.iter() {
+            if !Self::is_split_contained_in_search_time_range(split, &request) {
+                continue;
             }
-            CanSplitDoBetter::SplitTimestampHigher(_)
-            | CanSplitDoBetter::FindTraceIdsAggregation(_) => {
-                splits.sort_unstable_by_key(|split| std::cmp::Reverse(split.timestamp_end()))
+
+            partial_sum += split.num_docs;
+
+            if partial_sum >= num_requested_docs {
+                return Some(min_required_splits + 1);
             }
-            CanSplitDoBetter::SplitTimestampLower(_) => {
-                splits.sort_unstable_by_key(|split| split.timestamp_start())
+
+            min_required_splits += 1;
+        }
+
+        None
+    }
+
+    fn optimize_split_id_higher(
+        &self,
+        request: Arc<SearchRequest>,
+        mut splits: Vec<SplitIdAndFooterOffsets>,
+    ) -> Result<Vec<(SplitIdAndFooterOffsets, SearchRequest)>, SearchError> {
+        splits.sort_unstable_by(|a, b| b.split_id.cmp(&a.split_id));
+
+        if !is_simple_all_query(&request) {
+            // no optimization opportunity here.
+            return Ok(Self::to_splits_with_request(splits, request));
+        }
+
+        let Some(min_required_splits) = Self::get_min_required_splits(&splits, &request) else {
+            // not enough splits contained in time range.
+            return Ok(Self::to_splits_with_request(splits, request));
+        };
+
+        let mut split_with_req = Self::to_splits_with_request(splits, request);
+
+        // In this case there is no sort order, we order by split id.
+        // If the the first split has enough documents, we can convert the other queries to
+        // count only queries.
+        for (_split, request) in split_with_req.iter_mut().skip(min_required_splits) {
+            disable_search_request_hits(request);
+        }
+
+        Ok(split_with_req)
+    }
+
+    fn optimize_split_timestamp_higher(
+        &self,
+        request: Arc<SearchRequest>,
+        mut splits: Vec<SplitIdAndFooterOffsets>,
+    ) -> Result<Vec<(SplitIdAndFooterOffsets, SearchRequest)>, SearchError> {
+        splits.sort_unstable_by_key(|split| {
+            let contained = Self::is_split_contained_in_search_time_range(split, &request);
+            (!contained, std::cmp::Reverse(split.timestamp_end()))
```

Comment:
It is really hard to know here if contained splits come first or second (the ordering of boolean is not obvious, the `!` negation doesn't help 😄 ). Could we maybe explicitly map true/false to something that is more clearly ordered:
```
let contained_first = match Self::is_split_contained_in_search_time_range(split, &request) {
  true => 0u8,
  false => 1u8,
};
(contained_first, std::cmp::Reverse(split.timestamp_end()))
```

Reply:
MORE IMPORTANT: now that I think about it, I thing it's not good idea to have contained splits processed first in the general case (`!is_simple_all_query` or get_min_required_splits().is_none()`). You miss an optimization opportunity, because once you'll have processed them, you will still need to check the splits overlapping the higher end of the range for better matches.

Reply:
hmm, we can maybe instead of sorting by contained, keep the same sort mechanism as it was.

but still keep the new implementation of `get_min_required_splits`, meaning we won't optimize the first few splits which might not be contained in the search query's range.

Reply:
made the change to not sort, tell me what you think.